lib/mayaUsd/ufe/UsdStageMap.cpp

```diff
@@ -230,9 +230,21 @@ UsdStageMap::StageSet UsdStageMap::allStages()
 {
     rebuildIfDirty();
 
-    StageSet stages;
+    // Note: stage() calls proxyShape() which can modify fPathToObject,
+    //       so we cannot call stage within a loop on fPathToObject.
+    //
+    //       So, we first cache all UFE paths and then loop over them
+    //       calling stage().
+
+    std::vector<Ufe::Path> paths;
+    paths.reserve(fPathToObject.size());
     for (const auto& pair : fPathToObject) {
-        stages.insert(stage(pair.first));
+        paths.emplace_back(pair.first);
+    }
+
+    StageSet stages;
+    for (const auto& path : paths) {
+        stages.insert(stage(path));
```

Comment:
I think this code will append an empty/invalid Stage to the set when processing a deleted stage (but only once since all deleted stages will return the same hash). Would it be better to test that the stage is valid before adding it to the StageSet? The function is called `allStages`.

Reply:
I had the same reflection, but the old code did not check the validity either. The stage map holds weak pointers, so the validity can change at any moment anyway...